DEVELOPMENT_GUIDE.md

```diff
@@ -0,0 +1,211 @@
+# Development Guide
+
+The following document will be a reference guide to coding standards of cp-ansible. It is written to be idempotent, meaning whatever you set in your inventory file, the playbook and roles should provision on your hosts, and can be run multiple times without causing unnecessary changes to those hosts.
+
+#### Table Of Contents
+
+[Roles](#roles)
+
+[Variables](#variables)
+
+[Component Properties](#component-properties)
+
+[Custom Filters](#custom-filters)
+
+[Testing](#testing)
+
+[Branching Model](#branding-model)
+
+## Roles
+Each Confluent component has its own role, with the name `confluent.<component_name>`. Within that role, `main.yml` is the entrypoint of all tasks run when the role is invoked. Here are a few tasks and coding standards associated with them:
+
+```
+- name: Write Service Overrides	(1)
+  template:	(2)
+    src: override.conf.j2
+    dest: "{{ kafka_broker.systemd_override }}"	(3)
+    mode: 0640	(4)
+    owner: "{{kafka_broker_user}}"	(5)
+    group: "{{kafka_broker_group}}"
+  notify: restart kafka	(6)
+```
+
+1. A name clearly defining what the task accomplishes, using capital letters
+2. Uses an idempotent ansible module whenever possible
+3. Make use of variables instead of hard coding paths
+4. For file creation use 0640 permission, for directory creation use 0750 permission. There are some exceptions, but be sure to secure files.
+5. Proper ownership set
+6. Trigger component restart handler when necessary
+
+```
+- name: Register Cluster
+  include_tasks: register_cluster.yml	(1)
+  when: kafka_connect_cluster_name|length > 0	(2)
+```
+
+1. Put sets of related tasks into their own task file and include them (not import)
+2. Use conditionals to skip entire sets of tasks, or singular tasks
+
+```
+- name: Import the CA cert into the Keystore
+  shell: |
+    keytool -noprompt -keystore {{keystore_path}} \
+      -storetype pkcs12 \
+      -alias CARoot \
+      -import -file {{ca_cert_path}} \
+      -storepass {{keystore_storepass}}
+  no_log: "{{mask_secrets|bool}}"	(1)
+```
+
+1. Passwords are not logged
+
+## Variables
+
+Variables are defined in multiple places within cp-ansible. Each role will have its own variables defined in `defaults/main.yml`. These variables can be used by tasks within that role, but not tasks from other roles. For variables that must be scoped to multiple roles there is the confluent.variables role. This role is set as a dependency for most of our roles:
+
```

Comment:
The same could be done to shared component variables that should not be modifiable by users. Can create `vars/main` directory and define the variable per file - according to which component "owns" the variable (even shared). This separation could also improve the doc-generation, since the variables will be organized per component (file).

Reply:
oh dang is that possible? so there'd be
```
confluent.variables/
  defaults/main/
    kafka_broker.yml
    zookeeper.yml
```
??

Looks like yes: https://stackoverflow.com/a/65777944/5352328

That sounds pretty good to me. The thing is this doc is valid as of today, so if we make that change then well just have to change this doc. We can track that in an GH issue

Reply:
Sounds like a plan! I just wanted to mention it. 😄